kernel/interrupts/src/aarch64/mod.rs

```diff
@@ -0,0 +1,416 @@
+
+        // unmask the interrupt
+        // enable the timer
+        CNTP_CTL_EL0.write(
+              CNTP_CTL_EL0::IMASK.val(0)
+            + CNTP_CTL_EL0::ENABLE.val(1)
+        );
+
+        /* DEBUGGING CODE
+
+        log::info!("timer counter: {:?}", CNTPCT_EL0.get());
+        log::info!("timer enabled: {:?}",  CNTP_CTL_EL0.read(CNTP_CTL_EL0::ENABLE));
+        log::info!("timer IMASK: {:?}",   CNTP_CTL_EL0.read(CNTP_CTL_EL0::IMASK));
+        log::info!("timer status: {:?}", CNTP_CTL_EL0.read(CNTP_CTL_EL0::ISTATUS));
+
+        */
+
+        Ok(())
+}
+
+/// Registers an interrupt handler at the given IRQ interrupt number.
+///
+/// The function fails if the interrupt number is reserved or is already in use.
+///
+/// # Arguments 
+/// * `irq_num`: the interrupt (IRQ vector) that is being requested.
+/// * `func`: the handler to be registered, which will be invoked when the interrupt occurs.
+///
+/// # Return
+/// * `Ok(())` if successfully registered, or
+/// * `Err(existing_handler_address)` if the given `irq_num` was already in use.
+pub fn register_interrupt(irq_num: IntNumber, func: HandlerFunc) -> Result<(), *const HandlerFunc> {
+    let mut handlers = IRQ_HANDLERS.write();
+    let index = irq_num as usize;
+
+    let value = handlers[index] as *const HandlerFunc;
+    let default = default_irq_handler as *const HandlerFunc;
+
+    if value == default {
+        handlers[index] = func;
+        Ok(())
+    } else {
+        error!("register_interrupt: the requested interrupt IRQ {} was already in use", irq_num);
+        Err(value)
+    }
+}
+
+/// Deregisters an interrupt handler, making it available to the rest of the system again.
+///
+/// As a sanity/safety check, the caller must provide the `interrupt_handler`
+/// that is currently registered for the given IRQ `interrupt_num`.
+/// This function returns an error if the currently-registered handler does not match 'func'.
+///
+/// # Arguments
+/// * `interrupt_num`: the IRQ that needs to be deregistered
+/// * `func`: the handler that should currently be stored for 'interrupt_num'
+pub fn deregister_interrupt(irq_num: IntNumber, func: HandlerFunc) -> Result<(), *const HandlerFunc> {
+    let mut handlers = IRQ_HANDLERS.write();
+    let index = irq_num as usize;
+
+    let value = handlers[index] as *const HandlerFunc;
+    let func = func as *const HandlerFunc;
+
+    if value == func {
+        handlers[index] = default_irq_handler;
+        Ok(())
+    } else {
+        error!("deregister_interrupt: Cannot free interrupt due to incorrect handler function");
+        Err(value)
+    }
+}
+
+/// Send an "end of interrupt" signal, notifying the interrupt chip that
+/// the given interrupt request `irq` has been serviced.
+pub fn eoi(irq_num: IntNumber) {
+    let mut gic = GIC.lock();
+    let gic = gic.as_mut().expect("GIC is uninitialized");
+    gic.end_of_interrupt(irq_num);
+}
+
+#[rustfmt::skip]
+impl fmt::Display for SpsrEL1 {
```

Comment:
These 3 display blocks should implement `Debug`, not `Display`. This is far too much inner details for a `Display`impl, imo.

Reply:
Converted it to Debug.